src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+class WP_HTML_Processor extends WP_HTML_Tag_Processor {
+	public $fully_supported_input = null;
+	public $open_elements = array();
+
+	public function ensure_support() {
+		if ( null !== $this->fully_supported_input ) {
+			return $this->fully_supported_input;
+		}
+
+		$stack = array();
+
+		$p = new WP_HTML_Tag_Processor( $this->html );
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			$tag_name = $p->get_tag();
+
+			if ( ! $p->is_tag_closer() ) {
+				$element = WP_HTML_Spec::element_info( $tag_name );
+
+				$self_closes = $element::is_void || ( ! $element::is_html && $p->has_self_closing_flag() );
+				if ( ! $self_closes ) {
+					$stack[] = $tag_name;
+				}
+			} else {
+				if ( end( $stack ) === $tag_name ) {
+					array_pop( $stack );
+					continue;
+				}
+
+				$this->fully_supported_input = false;
+				return false;
+			}
+		}
+
+		$this->fully_supported_input = 0 === count( $stack );
+
+		return $this->fully_supported_input;
+	}
+
+	public function next_tag( $query = null ) {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		if ( 0 < count( $this->open_elements ) ) {
+			$element = WP_HTML_Spec::element_info( end( $this->open_elements ) );
+			// @TODO: Handle self-closing HTML foreign elements: must convey self-closing flag on stack.
+			if ( $element::is_void ) {
+				array_pop( $this->open_elements );
+			}
+		}
+
+		if ( false === parent::next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			return false;
+		}
+
+		$tag_name = $this->get_tag();
+		$element = WP_HTML_Spec::element_info( $tag_name );
+
+		$self_closes = $element::is_void || ( ! $element::is_html && $this->has_self_closing_flag() );
+		if ( $self_closes ) {
+			$this->open_elements[] = $tag_name;
+			return true;
+		}
+
+		if ( $this->is_tag_closer() ) {
+			array_pop( $this->open_elements );
+		} else {
+			$this->open_elements[] = $tag_name;
+		}
+
+		return true;
+	}
+
+	public function next_sibling() {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		$starting_depth = count( $this->open_elements );
+
+		while ( $this->next_tag() ) {
+			$current_depth = count( $this->open_elements );
+
+			if ( ! $this->is_tag_closer() && $current_depth === $starting_depth ) {
+				return true;
+			}
+
+			if ( ! $this->is_tag_closer() && $current_depth < $starting_depth ) {
+				return false;
+			}
+		}
+
+		return false;
+	}
+
+	public function next_child() {
+		if ( false === $this->fully_supported_input || false === $this->ensure_support() ) {
+			return false;
+		}
+
+		$starting_depth = count( $this->open_elements );
+
+		while ( $this->next_tag() ) {
```

Comment:
We should bale out if between the $starting_depth and the child we see anything with `$current_depth <= $starting_depth` as that would imply the initial tag was closed

Reply:
given the `ensure_support()` function that runs I don't think it's possible for this to happen, but I think I can defer some of the support function. I've added a test to make sure, though I haven't yet added any non-HTML elements in the tests.